src/Microsoft.VisualStudio.Editor.Razor/VisualStudioRazorParser.cs

```diff
@@ -2,266 +2,26 @@
-                if (_partialParser != null && _parser.IsIdle)
-                {
-                    result = _partialParser.Parse(change);
-                }
-            }
-
-            // If partial parsing failed or there were outstanding parser tasks, start a full reparse
-            if ((result & PartialParseResultInternal.Rejected) == PartialParseResultInternal.Rejected)
-            {
-                _parser.QueueChange(change, snapshot);
-            }
-
-            if ((result & PartialParseResultInternal.Provisional) == PartialParseResultInternal.Provisional)
-            {
-                StartIdleTimer();
-            }
-        }
-
-        private void OnIdle(object state)
-        {
-            _dispatcher.AssertForegroundThread();
-
-            var documentTracker = _documentTrackerFactory.GetTracker(_textBuffer);
-
-            if (documentTracker == null)
-            {
-                Debug.Fail("Document tracker should never be null when checking idle state.");
-                return;
-            }
-
-            foreach (var textView in documentTracker.TextViews)
-            {
-                if (_completionBroker.IsCompletionActive(textView))
-                {
-                    // Completion list is still active, need to re-start timer.
-                    StartIdleTimer();
-                    return;
-                }
-            }
-
-            Reparse();
-        }
-
-        private async void Timer_Tick(object state)
-        {
-            try
-            {
-                _dispatcher.AssertBackgroundThread();
-
-                StopIdleTimer();
-
-                // We need to get back to the UI thread to properly check if a completion is active.
-                await Task.Factory.StartNew(OnIdle, null, CancellationToken.None, TaskCreationOptions.None, _dispatcher.ForegroundScheduler);
-            }
-            catch (Exception ex)
-            {
-                // This is something totally unexpected, let's just send it over to the workspace.
-                await Task.Factory.StartNew(() => _errorReporter.ReportError(ex), CancellationToken.None, TaskCreationOptions.None, _dispatcher.ForegroundScheduler);
-            }
-        }
-
-        private void OnResultsReady(object sender, DocumentStructureChangedEventArgs args)
-        {
-            _dispatcher.AssertBackgroundThread();
-
-            if (DocumentStructureChanged != null)
-            {
-                if (args.Snapshot != _textBuffer.CurrentSnapshot)
-                {
-                    // A different text change is being parsed.
-                    return;
-                }
-
-                CodeDocument = args.CodeDocument;
-                Snapshot = args.Snapshot;
-                _partialParser = new RazorSyntaxTreePartialParser(CodeDocument.GetSyntaxTree());
-                DocumentStructureChanged(this, args);
-            }
-        }
+        public abstract Task ReparseAsync();
```

Comment:
Does this task complete when the parse is completed? Or when then the parse is enqued?

Reply:
https://github.com/aspnet/Razor/pull/1722#discussion_r145479745